### PR TITLE
refactor: replace jsoncpp with rapidjson.

### PR DESCRIPTION
**Pull request by soodoshll**
- check format when reading config files (using try&catch)
- some changes of the extern directory and CMakeLists.txt
- eliminate warnings
- fix a bug in UI
- make lane changing more natural

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -42,8 +38,16 @@ endif()
 
 add_subdirectory(extern/pybind11)
 pybind11_add_module(${PROJECT_NAME} MODULE src/simulator.cpp)
+
+if(NOT EXISTS "${PROJECT_SOURCE_DIR}/extern/rapidjson/include")
+    message(FATAL_ERROR "The submodules were not downloaded! GIT_SUBMODULE was turned off or failed. Please update submodules and try again.")
+endif()
+include_directories(extern/rapidjson/include)
+
+add_subdirectory(src)
```
Comment (zhc134) [inline CMakeLists.txt]: why do you put `add_subdirectory(src)` here?
does is still work since you have put this line after 
`pybind11_add_module(${PROJECT_NAME} MODULE src/simulator.cpp)`

File: src/engine/engine.cpp [hunk 2]
```diff
@@ -1,6 +1,8 @@
 #include "engine/engine.h"
 #include "utility/utility.h"
-#include "json/json.h"
+#include "rapidjson/filereadstream.h"
+#include "rapidjson/filewritestream.h"
+#include <rapidjson/writer.h>
```
Comment (zhc134) [inline src/engine/engine.cpp]: why not use `"rapidjson/writer.h"`

File: src/roadnet/roadnet.h [hunk 3]
```diff
@@ -316,7 +316,7 @@ namespace CityFlow {
 
         bool canEnter(const Vehicle *vehicle) const;
 
-        int getLaneIndex() const { return this->laneIndex; }
+        std::size_t getLaneIndex() const { return this->laneIndex; }
```
Comment (zhc134) [inline src/roadnet/roadnet.h]: why use std::size_t instead of size_t?

File: src/roadnet/roadnet.h [hunk 4]
```diff
@@ -365,7 +365,7 @@ namespace CityFlow {
 
         std::vector<Segment> &getSegments() { return segments; }
 
-        size_t getSegmentNum() { return segments.size(); }
+        size_t getSegmentNum() const{ return segments.size(); }
```
Comment (zhc134) [inline src/roadnet/roadnet.h]: space after const

File: src/utility/utility.cpp [hunk 5]
```diff
@@ -84,4 +85,89 @@ namespace CityFlow {
 		std::shuffle(randoms.begin(), randoms.end(), *rnd);
 		return randoms;
 	}
+
+    int getJsonMemberInt(const std::string &name, const rapidjson::Value &object) {
+        assert(object.IsObject());
+	    auto iter = object.FindMember(name.c_str());
```
Comment (zhc134) [inline src/utility/utility.cpp]: strange indentation

File: src/utility/utility.cpp [hunk 6]
```diff
@@ -84,4 +85,89 @@ namespace CityFlow {
 		std::shuffle(randoms.begin(), randoms.end(), *rnd);
 		return randoms;
 	}
+
+    int getJsonMemberInt(const std::string &name, const rapidjson::Value &object) {
+        assert(object.IsObject());
```
Comment (Hzfengsy) [inline src/utility/utility.cpp]: use 4 spaces instead of tab

File: src/utility/utility.h [hunk 7]
```diff
@@ -75,6 +76,37 @@ namespace CityFlow {
...
+    };
+
+    class jsonTypeError: public jsonFormatError  {
+    private:
+        std::string info;
+    public:
+        jsonTypeError(const std::string &name, const std::string &type) :
+            jsonFormatError(name + ": expected type " + type){}
+    };
+	int getJsonMemberInt(const std::string &name, const rapidjson::Value &object);
```
Comment (zhc134) [inline src/utility/utility.h]: strange indentation

File: src/vehicle/lanechange.cpp [hunk 8]
```diff
@@ -13,7 +13,7 @@ namespace CityFlow{
         return signalSend ? signalSend->target : (Lane *)vehicle->getCurDrivable();
     }
 
-    bool LaneChange::planChange() {
+    bool LaneChange::planChange() const{
```
Comment (zhc134) [inline src/vehicle/lanechange.cpp]: space after const

File: src/vehicle/lanechange.cpp [hunk 9]
```diff
@@ -142,28 +139,28 @@ namespace CityFlow{
...
         if (vehicle->getCurDrivable()->isLane()) {
             Lane * curLane = (Lane *)vehicle->getCurDrivable();
 
             if (curLane->getLength() - vehicle->getDistance() < 30) return;
             double curEst = vehicle->getGap();
             double outerEst = 0;
-            double expectedGap = vehicle->getLen() + interval * vehicle->getSpeed();
-            Router &router = vehicle->controllerInfo.router;
+            double expectedGap = 2 * vehicle->getLen() + 4 * interval * vehicle->getMaxSpeed();
+            if (vehicle->getGap() > expectedGap || vehicle->getGap() < 1.5 * vehicle->getLen()) return;
```
Comment (zhc134) [inline src/vehicle/lanechange.cpp]: what's the reason behind these changes?

File: src/vehicle/lanechange.cpp [hunk 10]
```diff
@@ -199,15 +196,15 @@ namespace CityFlow{
...
         return targetFollower ? targetFollower->getMinBrakeDistance() : 0;
     }
 
-    double SimpleLaneChange::safeGapAfter() {
+    double SimpleLaneChange::safeGapAfter() const {
         return vehicle->getMinBrakeDistance();
     }
 
-    double SimpleLaneChange::estimateGap(Lane *lane) {
+    double SimpleLaneChange::estimateGap(const Lane *lane) const{
```
Comment (zhc134) [inline src/vehicle/lanechange.cpp]: space after const

File: src/engine/engine.cpp (hunk 2 above)
Comment (Hzfengsy) [inline src/engine/engine.cpp]: we should use either `#include<>` or `#include""`, but not both of them

File: src/engine/engine.cpp [hunk 11]
```diff
@@ -37,47 +40,55 @@ namespace CityFlow {
...
-        rlTrafficLight = root["rlTrafficLight"].asBool();
-
-        if (root.isMember("laneChange"))
-            laneChange = root["laneChange"].asBool();
-        else
-            laneChange = false;
-
-        int seed = root["seed"].asInt();
-        rnd.seed(seed);
+        char readBuffer[65536];
```
Comment (Hzfengsy) [inline src/engine/engine.cpp]: I am not sure what will happen if the file length is larger than 65536

File: src/roadnet/roadnet.h (hunk 3 above)
Comment (Hzfengsy) [inline src/roadnet/roadnet.h]: Do we need use `std:: size_t` here? can we only use `size_t`?

File: src/engine/engine.cpp (hunk 11 above)
Comment (zhc134) [reply src/engine/engine.cpp]: this is just a buffer for stream reader, so it should not be a concern? @soodoshll please verify

File: src/utility/utility.cpp [hunk 12]
```diff
@@ -84,4 +85,89 @@ namespace CityFlow {
...
+        return iter->value.GetDouble();
+    }
+
+    bool getJsonMemberBool(const std::string &name, const rapidjson::Value &object, bool default_value) {
+        assert(object.IsObject());
+        auto iter = object.FindMember(name.c_str());
+        if (iter == object.MemberEnd() || !iter->value.IsBool())
+            return default_value;
+        return iter->value.GetBool();
+    }
```
Comment (zhc134) [inline src/utility/utility.cpp]: any suggestions on simplifying these bunch of functions? @Hzfengsy

File: src/engine/engine.cpp (hunk 11 above)
Comment (soodoshll) [reply src/engine/engine.cpp]: testcase is much larger than 64k and it seems no problem here

Comment (Hzfengsy) [reply src/engine/engine.cpp]: Since we don't know the effect of the buffer length, please change the buffer size to a const variable which is defined as a hyperparameter.

File: src/utility/utility.h (hunk 7 above)
Comment (Hzfengsy) [inline src/utility/utility.h]: use spaces instead of tab

File: src/utility/utility.h [hunk 13]
```diff
@@ -75,6 +76,37 @@ namespace CityFlow {
...
+	double getJsonMemberDouble(const std::string &name, const rapidjson::Value &object);
+    bool getJsonMemberBool(const std::string &name, const rapidjson::Value &object);
+    std::string getJsonMemberString(const std::string &name, const rapidjson::Value &object);
+
+    int getJsonMemberInt(const std::string &name, const rapidjson::Value &object, int default_value);
+    double getJsonMemberDouble(const std::string &name, const rapidjson::Value &object, double default_value);
+    bool getJsonMemberBool(const std::string &name, const rapidjson::Value &object, bool default_value);
```
Comment (Hzfengsy) [inline src/utility/utility.h]: Can we use template function here?

File: src/vehicle/lanechange.cpp [hunk 14]
```diff
@@ -142,28 +139,28 @@ namespace CityFlow{
...
-            Router &router = vehicle->controllerInfo.router;
+            double expectedGap = 2 * vehicle->getLen() + 4 * interval * vehicle->getMaxSpeed();
+            if (vehicle->getGap() > expectedGap || vehicle->getGap() < 1.5 * vehicle->getLen()) return;
 
+            Router &router = vehicle->controllerInfo.router;
             if (curLane->getLaneIndex() < curLane->getBelongRoad()->getLanes().size() - 1){
                 if (router.onLastRoad() || router.getNextDrivable(curLane->getOuterLane())) {
                     outerEst = estimateGap(curLane->getOuterLane());
-                    if (outerEst > curEst + expectedGap)
+                    if (outerEst > curEst + vehicle->getLen())
```
Comment (Hzfengsy) [inline src/vehicle/lanechange.cpp]: Please add necessary comments when submitting a PR if you want to change the car following model or other rules. (Adding the comments on Github rather than in the code). Later we need to add detailed comments about the rules we used in the code.

File: src/vehicle/router.cpp [hunk 15]
```diff
@@ -42,7 +42,7 @@ namespace CityFlow {
     }
 
     Drivable *Router::getNextDrivable(int i) const {
-        if (i < planned.size() && i >= 0) {
```
Comment (Hzfengsy) [inline src/vehicle/router.cpp]: What's the meaning when `i < 0`? If it makes no sense, please change `int i` to `size_t i`

File: src/vehicle/router.cpp [hunk 16]
```diff
@@ -186,17 +187,21 @@ namespace CityFlow {
                 auto iter = dis.find(adjRoad);
                 double newDis;
 
+                double avgDur;
```
Comment (Hzfengsy) [inline src/vehicle/router.cpp]: Why move the `avgDur` to outside? It seems to be used in one case.

File: src/vehicle/vehicle.cpp [hunk 17]
```diff
@@ -212,7 +212,6 @@ namespace CityFlow {
         // collision free
         double v = getNoCollisionSpeed(leader->getSpeed(), leader->getMaxNegAcc(), vehicleInfo.speed,
                                        vehicleInfo.maxNegAcc, controllerInfo.gap, interval, 0);
-        int cnt = 0;
 
 //        if (v == -100 || v < vehicleInfo.speed - vehicleInfo.maxNegAcc * interval - 1e-7 ) return -200;
```
Comment (Hzfengsy) [inline src/vehicle/vehicle.cpp]: Please remove the useless code. I know it is not your fault, but could you give us a hand? Thank you!

File: src/engine/engine.cpp (hunk 11 above)
Comment (zhc134) [reply src/engine/engine.cpp]: agree @soodoshll

**Comment by soodoshll** [comment]
I'm confused with the usage of "std::". We use "abs" without the prefix while using "std::ceil". Which one  is the standard?

File: src/roadnet/roadnet.cpp [hunk 18]
```diff
@@ -239,24 +235,24 @@ namespace CityFlow {
                 const auto &lightPhasesValue = getJsonMemberArray("lightphases", trafficLightValue);
                 for (const auto &lightPhaseValue : lightPhasesValue.GetArray()) {
                     if (!lightPhaseValue.IsObject())
-                        throw jsonTypeError("lightphase", "object");
+                        throw JsonTypeError("lightphase", "object");
                     LightPhase lightPhase;
-//                    lightPhase.phase = static_cast<unsigned int>(getJsonMemberInt("phase", lightPhaseValue));
-                    lightPhase.time = getJsonMemberDouble("time", lightPhaseValue);
+//                    lightPhase.phase = static_cast<unsigned int>(getJsonMember<int>("phase", lightPhaseValue));
```
Comment (Hzfengsy) [inline src/roadnet/roadnet.cpp]: Please delete the useless code
